### ISSUE TITLE
Obtaining Hartree-Fock exchange energy density on grids

### ISSUE THREAD

**Issue by jc-umana-FI** [pyscf/pyscf#2011]
Hello pySCF team, I hope the new year is treating you well. 

I'm trying to obtain Hartree-Fock exchange energy densities on grids (as the title would suggest). While there was an issue ( #1485 ) posted and seemingly resolved a little over a year ago, there was no input from any pySCF devs and the issue was closed the same day that it was posted with no commentary from anyone other than the author. I've tried executing this method to no success, as it doesn't seem to return values that get us the energy when combined with rho and the grid weights (as you would be able to with exc in a Kohn-Sham calculation).

In short, I'm looking for an analogue to exc from KS, but in HF (an ex, if you will).

If there is any insight on how this may be accomplished, please let me know and I look forward to your responses!

**Comment by susilehtola** [comment]
$E_x$ does have an exact definition: it is the Hartree-Fock exchange energy. The correlation energy is defined as the difference of the exact energy and the exchange-only energy.

Note that the exchange energy **density** is ambiguous: if you have $\epsilon_x({\bf r})$ that integrates to $$E_x = \int \epsilon_x({\bf r}) {\rm d}^3 r,$$ you will find that any function $\epsilon_x'({\bf r})=\epsilon_x({\bf r})+f({\bf r})$ likewise gives the same exchange energy, as long as

 $$ \int f({\bf r}) {\rm d}^3 r=0 .$$

**Comment by sunqm** [comment]
I guess you try to compute $V(r') = \int \phi_i(r)\phi_j(r) \frac{1}{|r-r'|} \phi_j(r')\phi_j(r') dr$
```
mf = mol.RHF().run()
# analytically computes v_ij(r') = \int \chi_i(r) \chi_j(r) / |r-r'| dr
ijr = mol.intor('int1e_grids', grids_coords)
ao = mol.eval_gto(grids_coords)
occ_orb = mf.mo_coeff[:,:nocc]
mo = ao.dot(occ_orb)
hfx = np.einsum('pqr,pi,qj,rj,ri->r', ijr, occ_orb, occ_orb, mo, mo)
```

**Comment by verena-neufeld** [comment]
Hi @jc-umana-FI , thanks for your question. Has it been resolved now and can this issue be closed?
